src/core/dfn.js

```diff
@@ -86,32 +86,64 @@ function computeTypeAndExport(dfn, linkingText) {
       // First one wins
       type = [...dfn.classList].find(className => knownTypesMap.has(className));
       validateDefinition(linkingText, type, dfn);
-      shouldExport = true;
       break;
 
     // Internal slots: attributes+ methods (e.g., [[some words]](with, optional, arguments))
     case slotRegex.test(linkingText):
-      shouldExport = false;
       type = processAsInternalSlot(linkingText, dfn);
       break;
   }
 
-  // If the Editor explicitly asked for it to be exported, so let's export it.
-  if (dfn.classList.contains("export")) shouldExport = true;
-
-  // Get closest type from context
-  if (!type) {
+  // Derive closest type
+  if (!type && !dfn.matches("[data-dfn-type]")) {
     /** @type {HTMLElement} */
     const closestType = dfn.closest("[data-dfn-type]");
     type = closestType?.dataset.dfnType;
   }
+  // only if we have type and one wasn't explicitly given.
+  if (type && !dfn.dataset.dfnType) {
+    dfn.dataset.dfnType = type;
+  }
+  // Finally, addContractDefaults() will add the type to the dfn if it's not there.
+  // But other modules may end up adding a type (e.g., the WebIDL module)
+}
 
-  if (!dfn.dataset.dfnType && type) dfn.dataset.dfnType = type;
-  if (shouldExport && !dfn.hasAttribute("data-noexport")) {
-    dfn.dataset.export = "";
+// Deal with export/no export
+function computeExport(dfn) {
+  switch (true) {
+    // Error if we have both exports and no exports.
+    case dfn.matches(".export.no-export"): {
+      const msg = docLink`Declares both "${"[no-export]"}" and "${"[export]"}" css class.`;
+      const hint = "Please use only one.";
+      showError(msg, name, { elements: [dfn], hint });
+      break;
+    }
+
+    // No export wins
+    case isNoExport(dfn):
+      if (dfn.matches("[data-export]")) {
+        const msg = docLink`Declares ${"[no-export]"} CSS class, but also has a "${"[data-export]"}" attribute.`;
+        const hint = "Please chose only one.";
+        showError(msg, name, { elements: [dfn], hint });
+        delete dfn.dataset.export;
+      }
+      dfn.dataset.noexport = "";
+      break;
+
+    // If the author explicitly asked for it to be exported, so let's export it.
+    case dfn.matches(".export") && !dfn.matches("[data-no-export]"):
```

Comment:
I'm confused with the usage of `data-no-export` and `data-noexport`

Reply:
oops... this is why `data-noexport` should have never been.